### ISSUE TITLE
Segmentation fault in S_scan_heredoc()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14462]
Migrated from [rt.perl.org#123712](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123712) (status was 'resolved')

Searchable as RT123712$

**Comment by p5pRT** [comment]
### From thecrux@gmail.com

Sigsegv testcase&#8203;:

&nbsp;    $ echo \-n '/$a\[/\<\<' | perl
&nbsp;    Use of bare \<\< to mean \<\<"" is deprecated at \- line 1\.
&nbsp;    zsh&#8203;: done                echo \-n '/$a\[/\<\<' | 
&nbsp;    zsh&#8203;: segmentation fault  perl

\(gdb\) run
Program received signal SIGSEGV\, Segmentation fault\.
\_\_memcmp\_sse4\_1 \(\) at \.\./sysdeps/x86\_64/multiarch/memcmp\-sse4\.S&#8203;:949
949     \.\./sysdeps/x86\_64/multiarch/memcmp\-sse4\.S&#8203;: No such file or directory\.

\(gdb\) bt
\#0  \_\_memcmp\_sse4\_1 \(\) at \.\./sysdeps/x86\_64/multiarch/memcmp\-sse4\.S&#8203;:949
\#1  0x000000000058f2ce in S\_scan\_heredoc \(s=0x0\) at toke\.c&#8203;:9274
\#2  Perl\_yylex \(\) at toke\.c&#8203;:5891
\#3  0x00000000005bc185 in Perl\_yyparse \(gramtype=\<optimized out>\) at perly\.c&#8203;:322
\#4  0x00000000004e2d45 in S\_parse\_body \(xsinit=0x426dc0 \<xs\_init>\, env=0x0\) at perl\.c&#8203;:2273
\#5  perl\_parse \(my\_perl=\<optimized out>\, xsinit=0x426dc0 \<xs\_init>\, argc=\<optimized out>\, 
&nbsp;    argv=\<optimized out>\, env=0x0\) at perl\.c&#8203;:1607
\#6  0x00000000004269dc in main \(argc=2\, argv=0x7fffffffe498\, env=0x7fffffffe4b0\) at perlmain\.c&#8203;:114
\#7  0x00007ffff70d4ec5 in \_\_libc\_start\_main \(main=0x426870 \<main>\, argc=2\, argv=0x7fffffffe498\, 
&nbsp;    init=\<optimized out>\, fini=\<optimized out>\, rtld\_fini=\<optimized out>\, stack\_end=0x7fffffffe488\)
&nbsp;    at libc\-start\.c&#8203;:287
\#8  0x0000000000426cf3 in \_start \(\)

\(gdb\) frame 1
\(gdb\) list
9269            \}
9270            linestr = shared\->ls\_linestr;
9271            bufend = SvEND\(linestr\);
9272            d = s;
9273            while \(s \< bufend \- len \+ 1 &&
9274              memNE\(s\,PL\_tokenbuf\,len\) \) \{
9275                if \(\*s\+\+ == '\\n'\)
9276                    \+\+PL\_parser\->herelines;
9277            \}
9278            if \(s >= bufend \- len \+ 1\) \{

Crash reproduced with perl 5\.18\, 5\.20\, 5\.21\.8
Bug was found by afl fuzzer \(http&#8203;://lcamtuf\.coredump\.cx/afl/\)

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 01 13&#8203;:26&#8203;:13 2015\, crux wrote&#8203;:
> Sigsegv testcase&#8203;:
> 
> $ echo \-n '/$a\[/\<\<' | perl
> Use of bare \<\< to mean \<\<"" is deprecated at \- line 1\.
>  zsh&#8203;: done                echo \-n '/$a\[/\<\<' |
> zsh&#8203;: segmentation fault  perl
> 
> \(gdb\) run
> Program received signal SIGSEGV\, Segmentation fault\.

This is likely related to \#123617\, though it appears to be more recent\.  I’m running a bisect\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 01 14&#8203;:38&#8203;:39 2015\, sprout wrote&#8203;:
> On Sun Feb 01 13&#8203;:26&#8203;:13 2015\, crux wrote&#8203;:
> > Sigsegv testcase&#8203;:
> >
> > $ echo \-n '/$a\[/\<\<' | perl
> > Use of bare \<\< to mean \<\<"" is deprecated at \- line 1\.
> >  zsh&#8203;: done                echo \-n '/$a\[/\<\<' |
> > zsh&#8203;: segmentation fault  perl
> >
> > \(gdb\) run
> > Program received signal SIGSEGV\, Segmentation fault\.
> 
> This is likely related to \#123617\, though it appears to be more
> recent\.  I’m running a bisect\.

4efe39d21e072e88e12e308ed1f068461f8ef778 is the first bad commit
commit 4efe39d21e072e88e12e308ed1f068461f8ef778
Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
Date&#8203;:   Wed Aug 29 22&#8203;:07&#8203;:18 2012 \-0700

&nbsp;    toke\.c&#8203;:scan\_heredoc&#8203;: Merge similar code
&nbsp;    
&nbsp;    The code for looking in outer lexing scopes was mostly identical to
&nbsp;    the code for looking in PL\_linestr\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @hvds

On Sun Feb 01 14&#8203;:41&#8203;:10 2015\, sprout wrote&#8203;:
> On Sun Feb 01 14&#8203;:38&#8203;:39 2015\, sprout wrote&#8203;:

> > This is likely related to \#123617\, though it appears to be more
> > recent\.  I’m running a bisect\.
> 
> 4efe39d21e072e88e12e308ed1f068461f8ef778 is the first bad commit
> commit 4efe39d21e072e88e12e308ed1f068461f8ef778
> Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> Date&#8203;:   Wed Aug 29 22&#8203;:07&#8203;:18 2012 \-0700
> 
>     toke\.c&#8203;:scan\_heredoc&#8203;: Merge similar code
>     
>     The code for looking in outer lexing scopes was mostly identical to
>     the code for looking in PL\_linestr\.

The below is one aspect changed from the original\, and is enough to make it survive \(and pass tests\); I'm not sure why reality doesn't match the various comments though \(so that we're here with \!infile and \!PL\_lex\_inwhat\, and it isn't an eval\, and doesn't have a newline\)\, so I'm not sure how to fix those\.

Hugo

<details><summary>Inline Patch</summary>

```diff
diff --git a/toke.c b/toke.c
index 24b5ed0..13f30e7 100644
--- a/toke.c
+++ b/toke.c
@@ -9275,7 +9275,8 @@ S_scan_heredoc(pTHX_ char *s)
          }
        else {  /* eval */
            s = (char*)memchr((void*)s, '\n', PL_bufend - s);
-           assert(s);
+           if (!s)
+                s = PL_bufend;
        }
        linestr = shared->ls_linestr;
        bufend = SvEND(linestr);
```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Feb 10 10&#8203;:15&#8203;:07 2015\, hv wrote&#8203;:
> On Sun Feb 01 14&#8203;:41&#8203;:10 2015\, sprout wrote&#8203;:
> > On Sun Feb 01 14&#8203;:38&#8203;:39 2015\, sprout wrote&#8203;:
> 
> > > This is likely related to \#123617\, though it appears to be more
> > > recent\.  I’m running a bisect\.
> >
> > 4efe39d21e072e88e12e308ed1f068461f8ef778 is the first bad commit
> > commit 4efe39d21e072e88e12e308ed1f068461f8ef778
> > Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> > Date&#8203;:   Wed Aug 29 22&#8203;:07&#8203;:18 2012 \-0700
> >
> > toke\.c&#8203;:scan\_heredoc&#8203;: Merge similar code
> >
> > The code for looking in outer lexing scopes was mostly identical to
> > the code for looking in PL\_linestr\.
> 
> The below is one aspect changed from the original\, and is enough to
> make it survive \(and pass tests\); I'm not sure why reality doesn't
> match the various comments though \(so that we're here with \!infile and
> \!PL\_lex\_inwhat\, and it isn't an eval\, and doesn't have a newline\)\, so
> I'm not sure how to fix those\.

A call to skipspace that happens inside /$a\[/ tries to read the next line of the file\.  When EOF is reached PL\_rsfp is set to NULL\, which is why S\_scan\_heredoc thinks we are inside an eval\, which is not the case\.

A similar circumstance can arise with&#8203;:

perl \-e 'print q|/$a\[\<\<end\]/\+\<\<| \. "\\nend"'|\./miniperl

which does indeed make S\_scan\_heredoc think it is in an eval\, but the first line does end with \\n\, so the memchr returns something and the assertion does not fail\.

scan\_heredoc’s assumptions are reasonable\.  I think we need to fix that /$a\[/ bug\, which will unfortunately break my japhs\. :\-\(

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Feb 10 12&#8203;:38&#8203;:15 2015\, sprout wrote&#8203;:
> On Tue Feb 10 10&#8203;:15&#8203;:07 2015\, hv wrote&#8203;:
> > On Sun Feb 01 14&#8203;:41&#8203;:10 2015\, sprout wrote&#8203;:
> > > On Sun Feb 01 14&#8203;:38&#8203;:39 2015\, sprout wrote&#8203;:
> >
> > > > This is likely related to \#123617\, though it appears to be more
> > > > recent\.  I’m running a bisect\.
> > >
> > > 4efe39d21e072e88e12e308ed1f068461f8ef778 is the first bad commit
> > > commit 4efe39d21e072e88e12e308ed1f068461f8ef778
> > > Author&#8203;: Father Chrysostomos \<sprout@&#8203;cpan\.org>
> > > Date&#8203;:   Wed Aug 29 22&#8203;:07&#8203;:18 2012 \-0700
> > >
> > > toke\.c&#8203;:scan\_heredoc&#8203;: Merge similar code
> > >
> > > The code for looking in outer lexing scopes was mostly identical to
> > > the code for looking in PL\_linestr\.
> >
> > The below is one aspect changed from the original\, and is enough to
> > make it survive \(and pass tests\); I'm not sure why reality doesn't
> > match the various comments though \(so that we're here with \!infile
> > and
> > \!PL\_lex\_inwhat\, and it isn't an eval\, and doesn't have a newline\)\, so
> > I'm not sure how to fix those\.
> 
> A call to skipspace that happens inside /$a\[/ tries to read the next
> line of the file\.  When EOF is reached PL\_rsfp is set to NULL\, which
> is why S\_scan\_heredoc thinks we are inside an eval\, which is not the
> case\.
> 
> A similar circumstance can arise with&#8203;:
> 
> perl \-e 'print q|/$a\[\<\<end\]/\+\<\<| \. "\\nend"'|\./miniperl
> 
> which does indeed make S\_scan\_heredoc think it is in an eval\, but the
> first line does end with \\n\, so the memchr returns something and the
> assertion does not fail\.
> 
> scan\_heredoc’s assumptions are reasonable\.  I think we need to fix
> that /$a\[/ bug\, which will unfortunately break my japhs\. :\-\(

I have finally fixed this in e47d32d\.  What I said about skipspace was not correct\.  skipspace was doing the right thing\, but many other paths were calling lex\_read\_space or lex\_next\_chunk without the proper guards\.  But putting guards all over the place seemed the wrong approach\, so I modified lex\_next\_chunk instead\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun Feb 15 17&#8203;:39&#8203;:17 2015\, sprout wrote&#8203;:
> I have finally fixed this in e47d32d\.  What I said about skipspace was
> not correct\.  skipspace was doing the right thing\, but many other
> paths were calling lex\_read\_space or lex\_next\_chunk without the proper
> guards\.  But putting guards all over the place seemed the wrong
> approach\, so I modified lex\_next\_chunk instead\.

I followed up with another fix in d27f4b9\.  A similar case was still crashing\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks for submitting this ticket

The issue should be resolved with the release today of Perl v5\.22\, available at http&#8203;://www\.perl\.org/get\.html
If you find that the problem persists\, feel free to reopen this ticket

\-\- 
Karl Williamson for the Perl 5 porters team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'

**Comment by p5pRT** [comment]
### From @karenetheridge

On Mon Jun 01 20&#8203;:50&#8203;:31 2015\, khw wrote&#8203;:
> Thanks for submitting this ticket
> 
> The issue should be resolved with the release today of Perl v5\.22\,
> available at http&#8203;://www\.perl\.org/get\.html
> If you find that the problem persists\, feel free to reopen this ticket


For tracability\, this fix appears to also have been backported to perl 5\.20\.3\, so the resulting BBC bug https://rt.perl.org/Public/Bug/Display.html?id=123865 is also in 5\.20\.3 \(therefore Devel\-Declare's fixes are needed for 5\.20\.3 as well\)\.
